### PR TITLE
7.x wrlc 66

### PR DESCRIPTION
**Pull request by MorganDawe**

### CODE REVIEW

**Comment by willtp87** [comment]
Had a chat with Morgan on how to avoid the same issues we ran into with the lack of an object table.

File: includes/islandora_usage_stats.class.inc [hunk 1]
```diff
@@ -119,6 +119,33 @@ class IslandoraUsageStats {
   }
 
   /**
+   * Increases the top datastream download count by one in the database.
+   *
+   * Count is not increased if user's IP is on exclusion list.
```
Comment (willtp87) [inline includes/islandora_usage_stats.class.inc:124]: Should we be respecting the time limit too? --handled

File: includes/islandora_usage_stats.class.inc [hunk 2]
```diff
@@ -144,7 +171,77 @@ class IslandoraUsageStats {
...
+   * Retrieves the number of times any object datastream has been downloaded.
+   *
+   * @param string $dsid
+   *   The datastream id.
+   *
+   * @return array
+   *   Records affected by top downloads.
+   */
+  private function getTopDownloads($dsid) {
+    $result = db_query("SELECT downloads, time from {islandora_usage_stats_top_downloads} WHERE pid_id = :pid_id AND dsid = :dsid", array(
```
Comment (willtp87) [inline includes/islandora_usage_stats.class.inc]: Queries should be in db.inc as much as possible.

File: islandora_usage_stats.install [hunk 3]
```diff
@@ -99,6 +121,51 @@ function islandora_usage_stats_schema() {
...
+        'type' => 'int',
+        'not null' => TRUE,
+      ),
+      'dsid' => array(
+        'description' => 'Datastream ID of the object',
+        'type' => 'varchar',
+        'length' => 64,
+        'not null' => TRUE,
+      ),
+      'downloads' => array(
```
Comment (willtp87) [inline islandora_usage_stats.install:178]: It would make sense to go more granular and have this simply be a count.

Comment (willtp87) [inline islandora_usage_stats.install:178]: Is islandora_usage_stats_object_ds_access_log tracking the same things as islandora_usage_stats_object_ds_downloads in the same fashion as the objects views/log tables?

File: includes/islandora_usage_stats.class.inc [hunk 4]
```diff
@@ -144,7 +171,77 @@ class IslandoraUsageStats {
...
+   *   The number of total downloads
+   */
+  private function increaseTopObjectDSDownloadCounter($dsid) {
+    $current_info = $this->getTopDownloads($dsid);
+    if ($current_info['downloads']) {
+      // Increase view count by one.
+      $result = db_update('islandora_usage_stats_top_downloads')
+      ->expression('downloads', 'downloads + :downloads',
+        array(
+          ':downloads' => 1,
```
Comment (willtp87) [inline includes/islandora_usage_stats.class.inc]: Why not throw in the 1 explicitly.

File: islandora_usage_stats.module [hunk 5]
```diff
@@ -424,3 +531,39 @@ function islandora_usage_stats_islandora_object_alter(AbstractObject $object, &$
...
+    = 'islandora_usage_stats_track_downloads';
+}
+
+/**
+ * Menu alter replacement callback for islandora/search.
+ *
+ * @param string $term
+ *   The search term to track.
+ */
+function islandora_usage_stats_track_search_results($term) {
```
Comment (willtp87) [inline islandora_usage_stats.module]: Doesn't need to be in the .module.

File: islandora_usage_stats.module [hunk 6]
```diff
@@ -424,3 +531,39 @@ function islandora_usage_stats_islandora_object_alter(AbstractObject $object, &$
...
+  return $output;
+}
+
+/**
+ * Menu alter replacement callback for ds downloads.
+ *
+ * @param AbstractDatastream $datastream
+ *   The datastream being downloaded.
+ */
+function islandora_usage_stats_track_downloads($datastream) {
```
Comment (willtp87) [inline islandora_usage_stats.module]: Doesn't need to be in the .module.

Comment (willtp87) [inline islandora_usage_stats.module]: Missing type hinting

File: islandora_usage_stats.module [hunk 7]
```diff
@@ -424,3 +531,39 @@ function islandora_usage_stats_islandora_object_alter(AbstractObject $object, &$
...
+  $ds_name = $datastream->id;
+  $parent_id = $datastream->parent->id;
+  $track = new IslandoraUsageStats($parent_id);
+  $track->addTopDownload($ds_name);
+  islandora_view_datastream($datastream, TRUE);
```
Comment (willtp87) [inline islandora_usage_stats.module]: Should likely program to the interface and call the wrapper https://github.com/Islandora/islandora/blob/7.x/includes/datastream.inc#L14

**Comment by willtp87** [comment]
Any new displays of the DB data should be views.

File: islandora_usage_stats.module (hunk 5 above)
Comment (MorganDawe) [inline islandora_usage_stats.module]: Tried moving the new callbacks in to another file, and used module load include in the menu alter(). Functions were not found. Move menu alter hook into the callbacks.inc file, menu alter was not found, callbacks not executing. Leaving them in the module.

Comment (willtp87) [inline islandora_usage_stats.module]: Did you fail to update the file key?

File: includes/db.inc [hunk 8]
```diff
@@ -103,6 +103,194 @@ function islandora_usage_stats_update_object_label($pid, $label) {
...
+ *   Using REQUEST_TIME, constant available time as int.
+ * @param int $pid_id
+ *   The pid_id, or id key in islandora_usage_stats_objects.
+ * @param int $dsid
+ *   The ds_id, or id key in islandora_usage_stats_datastreams.
+ *
+ * @return array
+ *   The query result.
+ */
+function islandora_usage_stats_increment_top_object_ds_download_counter($time, $pid_id, $dsid) {
```
Comment (nigelgbanks) [inline includes/db.inc:140]: Unused parameter `$time`.

File: includes/db.inc [hunk 9]
```diff
@@ -103,6 +103,194 @@ function islandora_usage_stats_update_object_label($pid, $label) {
...
+
+/**
+ * Finds top downloads for given information.
+ *
+ * In addition to finding top downloads, this
+ * function will also create an entry in
+ * islandora_usage_stats_datastreams if it
+ * does not exist.
+ */
+function islandora_usage_stats_find_existing_downloads($pid_id, $dsid, $label) {
```
Comment (nigelgbanks) [inline includes/db.inc:189]: Side note, although I'm not recommending we change this now, it's usually best from a testing debugging standpoint to have functions that only do one function at a time. So one that finds and existing download or returns FALSE and a separate function for creating an entry in the table.

File: islandora_usage_stats.module [hunk 10]
```diff
@@ -176,15 +217,79 @@ function islandora_usage_stats_block_view($delta = '') {
...
+    case 'top_downloads':
+      if (user_access('view usage stats block')) {
+        module_load_include('inc', 'islandora_usage_stats', 'includes/db');
+        $result = islandora_usage_stats_get_top_downloads(
+          variable_get('islandora_usage_stats_top_downloads_block_display_rows', 5)
+        );
+        $items = array();
+        foreach ($result as $pid) {
+          // @XXX This object load can fail perms, or not exist.
+          $object = islandora_object_load($pid);
```
Comment (nigelgbanks) [inline islandora_usage_stats.module:236]: Should check and not render the link if that's the case.

**Pull request merged**